decavision/dataset_preparation/generate_pseudolabels.py

```diff
@@ -0,0 +1,159 @@
+        image = tf.cast(image, tf.uint8)
+        # Normalize [0, 1]
+        image = tf.image.convert_image_dtype(image, dtype=tf.float32)
+        image = image.numpy()
+        return image
+
+    def generate_pseudolabel_data(self, model_path,
+                                  train_data_path, unlabeled_path,
+                                  pseudo_data_path, save_confidences_csv=False,
+                                  plot_confidences=False, threshold=None):
+        """ Convert image and label to tfrecord example.
+            Arguments:
+                model_path (str): path to trained model
+                train_path (str): path to training data
+                unlabeled_path (str): path to unlabeled data
+                pseudo_path (str): path to store train data and pseudo data
+                save_confidences_csv (boolean): Whether to save CSV file with pseudo confidences, default is False.
+                plot_confidences (boolean): Whether to plot confidence graphs for raw confidences and per class confidences.
+                threshold (float): Discard images with prediction below this confidence, default is None.
+
+            Returns:
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(train_data_path, pseudo_data_path)
+
+        # Load model
+        model = load_model(model_path, compile=False)
+        print("Loaded model.")
+
+        # Make dictionary for classes and their index
+        class_names = sorted(os.listdir(train_data_path))
+        class_dict = {}
+        index = 0
+        for cat in class_names:
+            class_dict[cat] = index
+            index += 1
+
+        print("Generating pseudo labels...")
+        # Generate pseudo labels
+        unlabeled_image_paths = os.listdir(unlabeled_path)
+        print("There are {} unlabeled images.".format(
+            len(unlabeled_image_paths)))
+        raw_predictions_paths = []
+        raw_predictions = []  # single confidence value of predicted class
+        raw_predictions_all = []  # confidences for all classes
+        for path in tqdm(unlabeled_image_paths[:100]):
+            # Load the image
+            img = np.array(self._load_img(os.path.join(unlabeled_path, path),
+                                          target_size=(299, 299)))
+            # Make predictions using trained model
+            y_pred = model.predict(np.expand_dims(img, axis=0), verbose=0)
+            # Get class index
+            y = np.argmax(y_pred)
+            # Get probability score
+            proba = y_pred[0][y]
+
+            raw_predictions.append(proba)
+            raw_predictions_all.append(y_pred)
+            raw_predictions_paths.append(os.path.join(unlabeled_path, path))
+
+            # The results variable should be the same as the class category
```

Comment:
I would also split the moving of the images to a different method, again to avoid having to generate the pseudolabels every time you want to do something. I think the method to generate the pseudolabels could save the probabilities in a csv and then the method to move the images uses that csv to decide where to move the images.

Reply:
The csv can also be used by the methods that generate plots.

Reply:
Made a separate method.